### DISCUSSION TITLE
Transform 8Bpp Tiff image to 2Bpp images change white to level 1 gray

### DISCUSSION THREAD

**Discussion by DanKrt82** [ImageMagick/ImageMagick#3508] [Help] [answered]
Hi,

I'm using ImageMagick  on Ubuntu:

```
/$ identify -version
Version: ImageMagick 6.9.10-23 Q16 x86_64 20190101 https://imagemagick.org
Copyright: © 1999-2019 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC Modules OpenMP
Delegates (built-in): bzlib djvu fftw fontconfig freetype jbig jng jpeg lcms lqr ltdl lzma openexr pangocairo png tiff webp wmf x xml zlib
```

 and I'm developing a tool on python that, at the end of the elaboration use wand to change the Bpp of the file (suppose that I already open the image (np_array variable) and make all the modifications that we need before the save of the image:

 ```
       with wandImage(wandImage.from_array(np_array.astype(np.uint8))) as img:
            img.depth = 2
            img.white_threshold = Color.black
            img.type = 'grayscale'
            img.resolution = (dpiX, dpiY)
            img.compression = 'undefined'
            img.save(filename='out_image.tif')
```

That procedure save the file like I need, but instead the value "255" (the white pixel on the original image) are set to "85" (in the result).

There is some configuration to change that behavior? I use the same code on Windows 10 and I don't have that problem, the images are ok at the end of the elaboration.

Thank you in advance

**Comment by snibgo** [comment]
> That procedure save the file like I need, but instead the value "255" (the white pixel on the original image) are set to "85" (in the result).

Please show input and output images.

**Reply by DanKrt82** [reply]
The input image (for example):

![image](https://user-images.githubusercontent.com/12826813/113708143-26eff480-96e1-11eb-95e0-1783d5996d80.png)

Output image:

![image](https://user-images.githubusercontent.com/12826813/113708208-3bcc8800-96e1-11eb-8051-57875be517a3.png)

**Comment by snibgo** [comment]
Those images are different sizes, so your code did not make that output from that input.

Please link to the actual files used. You could zip them together, upload the zip, and link to the zip.

**Reply by DanKrt82** [reply]
Ok, there the images (the white one is the input, the gray is the output):

[Pass_205.zip](https://github.com/ImageMagick/ImageMagick/files/6264565/Pass_205.zip)

**Comment by snibgo** [comment]
Your output TIFF has one channel, grayscale, 2 bits per pixel, so possible pixel values are 0, 1, 2 or 3. Every pixel is either 0 or 1. 1 out of 3 is the same as 85 out of 255.

I can't reproduce that at the command line, with IM v7.0.8-64:
```
magick Pass_205_input.tif -depth 2 -white-threshold 100% -type grayscale x.tiff
```
This makes an output that is also one channel, grayscale, 2 bits per pixel, but every pixel is either 0 (black) or 3 (white).

What version of IM are you using?

Incidentally, the density of your input and output is strange: 1x400, with no units. Density is usually the same in each direction. This should have no impact on your problem.

**Reply by DanKrt82** [reply]
Sorry, I saw that my DPI setting are 1  and not 400 (or 800) but the result is the same using the code that I reply on my post.

I use that version: ImageMagick 6.9.10-23 Q16 x86_64 on Ubuntu 20.04 with Python3.9

**Comment by snibgo** [comment]
As a test, I suggest you save the image immediately after from_array:
```
      with wandImage(wandImage.from_array(np_array.astype(np.uint8))) as img:
           img.save(filename='test_out_image.tif')
           img.depth = 2
           img.white_threshold = Color.black
           img.type = 'grayscale'
           img.resolution = (dpiX, dpiY)
           img.compression = 'undefined'
           img.save(filename='out_image.tif')
```
Please zip and link to test_out_image.tif.

**Reply by DanKrt82** [reply]
If I save that immediately after from array I have a red image (like you can see on the output images). I don't understand why, but if I save that in RGB the 255 value is set for the red, I attach two images like you ask, but I still don't understand why if I run the same code on a windows environment all it's working like expected.

Thank you very much!

[out_image.zip](https://github.com/ImageMagick/ImageMagick/files/6265038/out_image.zip)

**Comment by snibgo** [comment, marked as answer]
Okay, so test_out_image has only red and black pixels. Converting that to grayscale makes gray (85/255) and black pixels. The code from "img.depth" is working as expected.

Why do you think the input should be black and white? Perhaps your array has only one channel, but you are reading it as if it has R, G and B channels.

**Reply by DanKrt82** [reply]
I don't read the data from an image, I use directly a numpy array (the np_array.astype(np.uint8), to be clear) that is the output of a series of calculations from another image and contains only 1 color (1 np bidimensional array) with only black and white pixels ((1024, 7614) is the shape). I need to replicate the same array on a new third dimension of the array (like an RGB?) before use it for the conversion? In my images I have a 8bpp images, whit only 4 values (like a 2bpp), I need to transform that in a 2bpp grayscale file (that's way I'm trying to use the transform functionality). It's not possible use directly a dibimensional array for a grayscale image?

Thank you

**Reply by DanKrt82** [reply]
I found the solution, using b'I' as channel_map is working like expected, the problem was related that the default numpy array expected is an RGB image:

 with wandImage(wandImage.from_array(np_array.astype(np.uint8), b'I')) as img:

Thank you for the support.
